File: Solutions/3610-find-x-sum-of-all-k-long-subarrays-i/solution.cpp
class Solution {
public:
    int x_sum(vector<int> arr, int x) {
        map<int, int> mp;
        for(auto el: arr) mp[el]++;
        cout << "HI!" << endl;
        for(auto el: arr) {
            cout << el << ' ';
        }
        cout << endl;
        // if(mp.size() < x) {
        //     int su = 0;
        //     for(auto el: arr) su += el;
        //     return su;
        // }
        vector<pair<int, int>> mmd;
        for(auto el: mp) mmd.push_back({el.second, el.first});
        sort(mmd.begin(), mmd.end());
        int n = mmd.size();
        int i = n-1;
        int su = 0;
        while(x > 0 && i >= 0) {
            su += (mmd[i].first * mmd[i].second);
            i--;
            x--;
        }
        return su;
    }

    vector<int> findXSum(vector<int>& nums, int k, int x) {
        vector<int> ans;
        int n = nums.size();
        for(int i = 0 ; i < n-k+1 ; i++) {
            vector<int> arr;
            for(int j = i ; j < i + k ; j++) arr.push_back(nums[j]);
            ans.push_back(x_sum(arr, x));
        }
        return ans;
    }
};

File: Solutions/3610-find-x-sum-of-all-k-long-subarrays-i/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @param {number} x
 * @return {number[]}
 */
var findXSum = function(nums, k, x) {
    const xsum = (arr, x) => {
        let freqs = {};
        for (let n of arr) freqs[n] = (freqs[n] || 0) + 1;

        const nums = Object.keys(freqs).sort((a, b) => freqs[b] - freqs[a] || +b - +a);
        let sum = 0;
        for (let i = 0; i < Math.min(x, nums.length); i++)
            sum += nums[i] * freqs[nums[i]];

        return sum;
    };

    const n = nums.length;
    let res = new Array(n - k + 1).fill(0);
    for (let i = 0; i < res.length; i++)
        res[i] = xsum(nums.slice(i, i + k), x);

    return res;
};
